docs/designs/keras_callback.md

```diff
@@ -0,0 +1,196 @@
+For detail, we can view the [source code](https://github.com/tensorflow/tensorflow/blob/cf7fcf164c9846502b21cebb7d3d5ccf6cb626e8/tensorflow/python/keras/callbacks.py#L189-L196
+) of `CallbackList` in Tensorflow.
+
+So, we can also use CallbackList to wrap the callbacks in ElasticDL.
+```python
+from tensorflow.python.keras.callbacks import CallbackList
+callbacks = callbacks()
+callback_list = CallbackList(callbacks)
+callback_list.on_batch_end(batch)
+```
+
+
+### Call the Methods of Callback using ParameterServerStrategy
+
+Using ParameterServerStrategy in ElasticDL, we will launch instances as master, worker and parameter server. 
+The master controls the start and end of job and epoch. The worker and parameter server is responsible to process each batch. So we need to call different methods on different instance according the call frequency.
+
+
+### Call Methods of Callback Per Batch
+
+The methods should be called per batch in `tf.keras.callbacks.Callback` is
+```python
+on_(train|test|predict)_batch_begin(self, batch, logs=None)
+on_(train|test|predict)_batch_end(self, batch, logs=None)
+```
+Each batch needs worker to calculate loss and gradient and parameter server to update gradients. So the worker and parameter server should call those methods per batch.
+
+For worker:
+```python
+batch_index = 0
+for batch in dataset:
+    callback_list.on_(train|test|predict)_batch_begin(self, batch=batch_index, logs=None)
+    process(batch)
+    callback_list.on_(train|test|predict)_batch_end(self, batch=batch_index, logs=None)
+    batch_index += 1
+```
+
+The parameter server only updates the gradients for each batch during training, so it doesn't need to 
+call the methods per batch for test and prediction. 
+```python
+def update_parameter_for_batch(gradients, model_version):
+    callback_list.on_train_batch_begin(self, batch=model_version) # model_version is the iteration number during training.
+    update_gradient(gradients)
+    callback_list.on_train_batch_end(self, batch=model_vesion)
+```
+
+The worker and parameter server will both call the `on_train_batch_begin|end` and per batch. So, the 2 methods will be called twice and can be called by either worker or parameter server. For example, the `on_train_batch_begin` of `LearningRateScheduler` can only be called by parameter server to adjust the learning rate for each batch. There are 2 proposals to solve the problem.
+
+1. Only the parameter server calls `on_train_batch_begin|end` and the worker doesn't. Because, we 
+have not observed a use case for callback to execute per batch by the worker.
+
+2. ElasticDL sets an argmuent `instance_role` in `params` of each callback and users get the `instance_role` in `on_train_batch_begin|end`. It may need users to understand why to get the `instance_role`.
+```python
+class LearningRateScheduler(tf.keras.callbacks.Callback):
+    def on_train_batch_begin(self, batch, logs=None):
+        instance_role = self.params.get("instance_role", None)
```

Comment:
Need to define `self.params` and outline what `__init__()` looks like

Reply:
Maybe we can add a method named `on_update_batch_begin` specified for PS.